misc/python/materialize/checks/materialized_views.py

```diff
@@ -65,3 +67,83 @@ def validate(self) -> Testdrive:
            """
             )
         )
+
+
+class MaterializedViewsAssertNotNull(Check):
+    def _can_run(self, e: Executor) -> bool:
+        # CREATE ROLE not compatible with older releases
+        return self.base_version >= MzVersion.parse("0.73.0-dev")
+
+    def initialize(self) -> Testdrive:
+        return Testdrive(
+            dedent(
+                """
+                > CREATE TABLE not_null_table (x INT, y INT, z INT);
+                > INSERT INTO not_null_table VALUES (NULL, 2, 3), (4, NULL, 6), (7, 8, NULL);
+                > CREATE MATERIALIZED VIEW not_null_view1 WITH (ASSERT NOT NULL x) AS SELECT * FROM not_null_table;
+            """
+            )
+        )
+
+    def manipulate(self) -> list[Testdrive]:
+        return [
+            Testdrive(dedent(s))
+            for s in [
+                """
+                > CREATE MATERIALIZED VIEW not_null_view2 WITH (ASSERT NOT NULL y) AS SELECT * FROM not_null_table;
+                > INSERT INTO not_null_table VALUES (NULL, 12, 13), (14, NULL, 16), (17, 18, NULL);
+                """,
+                """
+                > CREATE MATERIALIZED VIEW not_null_view3 WITH (ASSERT NOT NULL z) AS SELECT * FROM not_null_table;
+                > INSERT INTO not_null_table VALUES (NULL, 22, 23), (24, NULL, 26), (27, 28, NULL);
+                """,
+            ]
+        ]
+
+    def validate(self) -> Testdrive:
+        return Testdrive(
+            dedent(
+                """
+                ! SELECT * FROM not_null_view1
+                contains: column 1 must not be null
+
+                ! SELECT * FROM not_null_view2
+                contains: column 2 must not be null
+
+                ! SELECT * FROM not_null_view3
+                contains: column 3 must not be null
+
+                > DELETE FROM not_null_table WHERE x IS NULL;
+
+                > SELECT * FROM not_null_view1
+                4 <null> 6
+                7 8 <null>
+                14 <null> 16
+                17 18 <null>
+                24 <null> 26
+                27 28 <null>
+
+                > DELETE FROM not_null_table WHERE y IS NULL;
+
+                > SELECT * FROM not_null_view2
```

Comment:
Sorry, one small thing, can you add an `? EXPLAIN SELECT * FROM not_null_viewXXX WHERE x IS NOT NULL` here.  The `EXPLAIN` should be `constant()`, that is, the optimizer has properly used the nullability information to determine that the WHERE clause is now impossible to satisfy and therefore the entire query should return a constant empty result.

This will confirm that the nullability annotation is actually being used to make optimization decisions. The SLT file with the original PR did not seem to have any EXPLAINs to that effect.